File: windows/src/autostart.rs
//! Launch-at-login, toggled through the per-user `Run` registry key. The macOS
//! app uses `SMAppService`; the Windows equivalent is a value under
//! `HKCU\Software\Microsoft\Windows\CurrentVersion\Run`.
//!
//! We drive it with `reg.exe` rather than the Win32 registry API — it keeps the
//! code tiny and needs no extra bindings, consistent with how the session
//! detector shells out to `git`.

use std::path::{Path, PathBuf};

use crate::util::command;

const RUN_KEY: &str = r"HKCU\Software\Microsoft\Windows\CurrentVersion\Run";
const VALUE_NAME: &str = "AgentCord";

/// The command the Run key should launch: this executable, quoted. With no
/// arguments it starts in tray mode.
fn exe_command() -> String {
    let exe = std::env::current_exe()
        .ok()
        .and_then(|p| p.to_str().map(str::to_string))
        .unwrap_or_default();
    format!("\"{exe}\"")
}

fn current_exe_path() -> Option<PathBuf> {
    std::env::current_exe().ok()
}

/// True when the Run value exists and points at this executable.
pub fn is_enabled() -> bool {
    let Some(stored) = read_run_value() else {
        return false;
    };
    let Some(expected) = current_exe_path() else {
        return false;
    };
    paths_match(&stored, &expected)
}

fn read_run_value() -> Option<String> {
    let output = command("reg")
        .args(["query", RUN_KEY, "/v", VALUE_NAME])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let text = String::from_utf8_lossy(&output.stdout);
    for line in text.lines() {
        let line = line.trim();
        if !line.starts_with(VALUE_NAME) {
            continue;
        }
        let (_, rest) = line.split_once("REG_SZ")?;
        return Some(rest.trim().trim_matches('"').to_string());
    }
    None
}

fn paths_match(stored: &str, expected: &Path) -> bool {
    let stored_path = Path::new(stored.trim().trim_matches('"'));
    if stored_path == expected {
        return true;
    }
    match (stored_path.canonicalize(), expected.canonicalize()) {
        (Ok(a), Ok(b)) => a == b,
        _ => false,
    }
}

/// Add or remove the Run value. Returns whether the change succeeded.
pub fn set_enabled(enabled: bool) -> bool {
    let result = if enabled {
        command("reg")
            .args(["add", RUN_KEY, "/v", VALUE_NAME, "/t", "REG_SZ", "/d", &exe_command(), "/f"])
            .output()
    } else {
        command("reg")
            .args(["delete", RUN_KEY, "/v", VALUE_NAME, "/f"])
            .output()
    };
    result.map(|o| o.status.success()).unwrap_or(false)
}

File: windows/Autostart.cs
// Launch-at-login, toggled through the per-user Run registry key. The macOS
// app uses SMAppService; the Windows equivalent is a value under
// HKCU\Software\Microsoft\Windows\CurrentVersion\Run.
//
// Writing Run is not enough on Windows 10/11: Settings / Task Manager also
// store an enable/disable flag under Explorer\StartupApproved\Run. A leftover
// "disabled" blob (first byte 0x03) keeps Explorer from launching the app
// even when the Run value is present and the in-app switch looks on.

using System.IO;
using System.Windows.Forms;
using Microsoft.Win32;

namespace AgentCord;

public static class Autostart
{
    private const string RunKey = @"Software\Microsoft\Windows\CurrentVersion\Run";
    private const string ApprovedKey =
        @"Software\Microsoft\Windows\CurrentVersion\Explorer\StartupApproved\Run";
    private const string ValueName = "AgentCord";

    private static string ExePath => Environment.ProcessPath ?? Application.ExecutablePath;

    /// <summary>True when the Run value points at this executable and Windows
    /// has not disabled the startup entry.</summary>
    public static bool IsEnabled()
    {
        try
        {
            using var key = Registry.CurrentUser.OpenSubKey(RunKey);
            var stored = key?.GetValue(ValueName) as string;
            if (string.IsNullOrEmpty(stored)) return false;
            if (!string.Equals(
                Path.GetFullPath(stored.Trim().Trim('"')),
                Path.GetFullPath(ExePath),
                StringComparison.OrdinalIgnoreCase))
            {
                return false;
            }

            using var approved = Registry.CurrentUser.OpenSubKey(ApprovedKey);
            return IsStartupApproved(approved?.GetValue(ValueName) as byte[]);
        }
        catch
        {
            return false;
        }
    }

    /// <summary>Add or remove the Run value and the matching StartupApproved
    /// flag. Returns whether the change succeeded.</summary>
    public static bool SetEnabled(bool enabled)
    {
        try
        {
            using var key = Registry.CurrentUser.CreateSubKey(RunKey);
            using var approved = Registry.CurrentUser.CreateSubKey(ApprovedKey);
            if (enabled)
            {
                // Quoted, no arguments: the exe starts in tray mode.
                key.SetValue(ValueName, $"\"{ExePath}\"");
                approved.SetValue(ValueName, EnabledApprovedBlob(), RegistryValueKind.Binary);
            }
            else
            {
                key.DeleteValue(ValueName, throwOnMissingValue: false);
                approved.DeleteValue(ValueName, throwOnMissingValue: false);
            }
            return true;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>Missing blob means Explorer has not vetoed the Run entry.
    /// 0x02 / 0x06 = enabled; 0x03 / 0x07 = disabled in Settings.</summary>
    internal static bool IsStartupApproved(byte[]? blob)
    {
        if (blob is null || blob.Length == 0) return true;
        return blob[0] is 0x02 or 0x06;
    }

    private static byte[] EnabledApprovedBlob()
    {
        var blob = new byte[12];
        blob[0] = 0x02;
        var time = BitConverter.GetBytes(DateTime.UtcNow.ToFileTimeUtc());
        Buffer.BlockCopy(time, 0, blob, 4, 8);
        return blob;
    }
}
